From 5340ed564b980b9d372af6b7bf7ef48b0d284cbf Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Mon, 20 Oct 2025 11:08:25 +0200
Subject: [PATCH 457/553] QChar: fix Clang 21 -Wcharacter-conversion warnings
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Clang 21 started warning about all conversions between charNN_t types
(since fixed for char16_t → char32_t¹), and QChar appears to be the
only header in QtBase that is affected by it. While char16_t →
char32_t has already been fixed upstream, we ought to have our public
headers clean of even this overzealous warning, so fix them by adding
an explicit cast. I would have used 'char32_t{ucs}', but the warning
is so buggy it's not fixed by braced initialization (filed as upstream
bug report https://github.com/llvm/llvm-project/issues/164220).

Interestingly, the warning does not trigger, for the same code, in
constexpr function, but proactively "fix" those, too, so we don't need
to re-touch these functions later.

As a drive-by, remove the pointless inline keywords from these
functions to make horizontal space for the cast.

Done-with: Thiago Macieira <thiago.macieira@intel.com>
Task-number: QTBUG-141204
Pick-to: 6.8 6.5
Change-Id: I2a9e46ea4f327fb554418c4649d189a98e15ffae
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
(cherry picked from commit f49331342f1b02d75c6405e8e2cd6cca5395b275)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/text/qchar.h | 62 ++++++++++++++++++++--------------------
 1 file changed, 31 insertions(+), 31 deletions(-)

diff --git a/src/corelib/text/qchar.h b/src/corelib/text/qchar.h
index cda50ad535c..5fae355733a 100644
--- a/src/corelib/text/qchar.h
+++ b/src/corelib/text/qchar.h
@@ -499,26 +499,26 @@ public:
         Unicode_16_0,
     };
 
-    inline Category category() const noexcept { return QChar::category(ucs); }
-    inline Direction direction() const noexcept { return QChar::direction(ucs); }
-    inline JoiningType joiningType() const noexcept { return QChar::joiningType(ucs); }
-    inline unsigned char combiningClass() const noexcept { return QChar::combiningClass(ucs); }
+    Category category() const noexcept { return QChar::category(char32_t(ucs)); }
+    Direction direction() const noexcept { return QChar::direction(char32_t(ucs)); }
+    JoiningType joiningType() const noexcept { return QChar::joiningType(char32_t(ucs)); }
+    unsigned char combiningClass() const noexcept { return QChar::combiningClass(char32_t(ucs)); }
 
-    inline QChar mirroredChar() const noexcept { return QChar(QChar::mirroredChar(ucs)); }
-    inline bool hasMirrored() const noexcept { return QChar::hasMirrored(ucs); }
+    QChar mirroredChar() const noexcept { return QChar(QChar::mirroredChar(char32_t(ucs))); }
+    bool hasMirrored() const noexcept { return QChar::hasMirrored(char32_t(ucs)); }
 
     QString decomposition() const;
-    inline Decomposition decompositionTag() const noexcept { return QChar::decompositionTag(ucs); }
+    Decomposition decompositionTag() const noexcept { return QChar::decompositionTag(char32_t(ucs)); }
 
-    inline int digitValue() const noexcept { return QChar::digitValue(ucs); }
-    inline QChar toLower() const noexcept { return QChar(QChar::toLower(ucs)); }
-    inline QChar toUpper() const noexcept { return QChar(QChar::toUpper(ucs)); }
-    inline QChar toTitleCase() const noexcept { return QChar(QChar::toTitleCase(ucs)); }
-    inline QChar toCaseFolded() const noexcept { return QChar(QChar::toCaseFolded(ucs)); }
+    int digitValue() const noexcept { return QChar::digitValue(char32_t(ucs)); }
+    QChar toLower() const noexcept { return QChar(QChar::toLower(char32_t(ucs))); }
+    QChar toUpper() const noexcept { return QChar(QChar::toUpper(char32_t(ucs))); }
+    QChar toTitleCase() const noexcept { return QChar(QChar::toTitleCase(char32_t(ucs))); }
+    QChar toCaseFolded() const noexcept { return QChar(QChar::toCaseFolded(char32_t(ucs))); }
 
-    inline Script script() const noexcept { return QChar::script(ucs); }
+    Script script() const noexcept { return QChar::script(char32_t(ucs)); }
 
-    inline UnicodeVersion unicodeVersion() const noexcept { return QChar::unicodeVersion(ucs); }
+    UnicodeVersion unicodeVersion() const noexcept { return QChar::unicodeVersion(char32_t(ucs)); }
 
     constexpr inline char toLatin1() const noexcept { return ucs > 0xff ? '\0' : char(ucs); }
     constexpr inline char16_t unicode() const noexcept { return ucs; }
@@ -528,23 +528,23 @@ public:
 
     constexpr inline bool isNull() const noexcept { return ucs == 0; }
 
-    inline bool isPrint() const noexcept { return QChar::isPrint(ucs); }
-    constexpr inline bool isSpace() const noexcept { return QChar::isSpace(ucs); }
-    inline bool isMark() const noexcept { return QChar::isMark(ucs); }
-    inline bool isPunct() const noexcept { return QChar::isPunct(ucs); }
-    inline bool isSymbol() const noexcept { return QChar::isSymbol(ucs); }
-    constexpr inline bool isLetter() const noexcept { return QChar::isLetter(ucs); }
-    constexpr inline bool isNumber() const noexcept { return QChar::isNumber(ucs); }
-    constexpr inline bool isLetterOrNumber() const noexcept { return QChar::isLetterOrNumber(ucs); }
-    constexpr inline bool isDigit() const noexcept { return QChar::isDigit(ucs); }
-    constexpr inline bool isLower() const noexcept { return QChar::isLower(ucs); }
-    constexpr inline bool isUpper() const noexcept { return QChar::isUpper(ucs); }
-    constexpr inline bool isTitleCase() const noexcept { return QChar::isTitleCase(ucs); }
-
-    constexpr inline bool isNonCharacter() const noexcept { return QChar::isNonCharacter(ucs); }
-    constexpr inline bool isHighSurrogate() const noexcept { return QChar::isHighSurrogate(ucs); }
-    constexpr inline bool isLowSurrogate() const noexcept { return QChar::isLowSurrogate(ucs); }
-    constexpr inline bool isSurrogate() const noexcept { return QChar::isSurrogate(ucs); }
+    bool isPrint() const noexcept { return QChar::isPrint(char32_t(ucs)); }
+    constexpr bool isSpace() const noexcept { return QChar::isSpace(char32_t(ucs)); }
+    bool isMark() const noexcept { return QChar::isMark(char32_t(ucs)); }
+    bool isPunct() const noexcept { return QChar::isPunct(char32_t(ucs)); }
+    bool isSymbol() const noexcept { return QChar::isSymbol(char32_t(ucs)); }
+    constexpr bool isLetter() const noexcept { return QChar::isLetter(char32_t(ucs)); }
+    constexpr bool isNumber() const noexcept { return QChar::isNumber(char32_t(ucs)); }
+    constexpr bool isLetterOrNumber() const noexcept { return QChar::isLetterOrNumber(char32_t(ucs)); }
+    constexpr bool isDigit() const noexcept { return QChar::isDigit(char32_t(ucs)); }
+    constexpr bool isLower() const noexcept { return QChar::isLower(char32_t(ucs)); }
+    constexpr bool isUpper() const noexcept { return QChar::isUpper(char32_t(ucs)); }
+    constexpr bool isTitleCase() const noexcept { return QChar::isTitleCase(char32_t(ucs)); }
+
+    constexpr bool isNonCharacter() const noexcept { return QChar::isNonCharacter(char32_t(ucs)); }
+    constexpr bool isHighSurrogate() const noexcept { return QChar::isHighSurrogate(char32_t(ucs)); }
+    constexpr bool isLowSurrogate() const noexcept { return QChar::isLowSurrogate(char32_t(ucs)); }
+    constexpr bool isSurrogate() const noexcept { return QChar::isSurrogate(char32_t(ucs)); }
 
     constexpr inline uchar cell() const noexcept { return uchar(ucs & 0xff); }
     constexpr inline uchar row() const noexcept { return uchar((ucs>>8)&0xff); }
-- 
2.51.2

